onnxscript/ir/_external_data.py

```diff
@@ -51,3 +75,209 @@ def set_base_dir(graph: _core.Graph | _core.GraphView, base_dir: str | os.PathLi
+    """Method to capture information about a tensor that is to be stored as external data."""
+    tensor_size = tensor.nbytes
+    # Calculate updated offset and align tensors
+    current_offset = compute_new_offset(current_offset, tensor_size)
+    # Store offset and tensor size as ExternalDataInfo
+    external_data_info = ExternalDataInfo(
+        tensor.name,  # type: ignore[arg-type]
+        current_offset,
+        tensor_size,
+    )
+    return external_data_info
+
+
+def _save_external_data(
+    external_data_info: list[tuple[_protocols.TensorProtocol, ExternalDataInfo]],
+    file_path: str | os.PathLike,
+) -> None:
+    """
+    Writes tensor data to an external file according to information stored in ExternalDataInfo object.
+
+    Args:
+        external_data_info: A collection of external data information stored for each tensor to be written as external data.
+        file_path: Location to which external data is to be stored.
+    """
+    with open(file_path, "w+b") as data_file:
+        for tensor, tensor_info in external_data_info:
+            current_offset = tensor_info.offset
+            assert tensor is not None
+            raw_data = tensor.tobytes()
+            if current_offset is not None:
+                # Pad file to required offset if needed
+                file_size = data_file.tell()
+                if current_offset > file_size:
+                    data_file.write(b"\0" * (current_offset - file_size))
+            data_file.write(raw_data)
+
+
+def _convert_as_external_tensors(
+    external_data_info: list[tuple[_protocols.TensorProtocol, ExternalDataInfo]],
+    file_path: str | os.PathLike,
+) -> list[_core.ExternalTensor]:
+    """
+    Convert the tensors (stored within the values) written as external data to _core.ExternalTensor types.
+
+    Args:
+        external_data_info: A collection of external data information stored for each tensor to be written as external data.
+        file_path: Location to which external data is to be stored.
+    """
+    external_tensors: list[_core.ExternalTensor] = []
+    for tensor, tensor_info in external_data_info:
+        assert tensor is not None
+        external_tensor = _core.ExternalTensor(
+            file_path,
+            tensor_info.offset,
+            tensor_info.length,
+            tensor.dtype,  # type: ignore[arg-type]
+            shape=tensor.shape,  # type: ignore[arg-type]
+            name=tensor.name,  # type: ignore[arg-type]
+        )
+        external_tensors.append(external_tensor)
+    return external_tensors
+
+
+def convert_tensors_to_external(
+    tensors: list[_protocols.TensorProtocol],
+    abs_path: str | os.PathLike,
+) -> list[_core.ExternalTensor]:
+    """
+    This method takes in tensors to be converted to external tensors and returns the converted external tensors.
+
+    Args:
+        tensors: Tensors to be converted to external tensors
+        abs_path: Absolute path to which external data is to be stored.
```

Comment:
We need a base path and a rel path. Otherwise the data written the tensor proto is incorrect.